### ISSUE TITLE
End of Sequence Error when using tf.estimator with tf.data

### ISSUE THREAD

**Issue by mrezak** [tensorflow/tensorflow#19541]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux CentOS v7
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: 1.8
- **Python version**: 2.7
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 9.0 / 7
- **GPU model and memory**: Titan Xp, 12GB 
- **Exact command to reproduce**: 

### Describe the problem
I am using tf.estimator.train_and_evaluate and tf.data.Dataset to feed data to the estimator:

Input Data function:

```
    def data_fn(data_dict, batch_size, mode, num_epochs=10):
        dataset = {}
        if mode == tf.estimator.ModeKeys.TRAIN:
            dataset = tf.data.Dataset.from_tensor_slices(data_dict['train_data'].astype(np.float32))
            dataset = dataset.cache()
            dataset = dataset.shuffle(buffer_size= batch_size * 10).repeat(num_epochs).batch(batch_size)
        else:
            dataset = tf.data.Dataset.from_tensor_slices(data_dict['valid_data'].astype(np.float32))
            dataset = dataset.cache()
            dataset = dataset.batch(batch_size)

        iterator = dataset.make_one_shot_iterator()
        next_element = iterator.get_next()

    return next_element
```

Train Function:
```
def train_model(data):
    tf.logging.set_verbosity(tf.logging.INFO)
    config = tf.ConfigProto(allow_soft_placement=True,
                            log_device_placement=False)
    config.gpu_options.allow_growth = True
    run_config = tf.contrib.learn.RunConfig(
        save_checkpoints_steps=10,
        keep_checkpoint_max=10,
        session_config=config
    )

    train_input = lambda: data_fn(data, 100, tf.estimator.ModeKeys.TRAIN, num_epochs=1)
    eval_input = lambda: data_fn(data, 1000, tf.estimator.ModeKeys.EVAL)
    estimator = tf.estimator.Estimator(model_fn=model_fn, params=hps, config=run_config)
    train_spec = tf.estimator.TrainSpec(train_input, max_steps=100)
    eval_spec = tf.estimator.EvalSpec(eval_input,
                                      steps=None,
                                      throttle_secs = 30)

    tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)
```

The training goes fine, but when it comes to evaluation I get this error:
`OutOfRangeError (see above for traceback): End of sequence `

If I don't use Dataset.batch on evaluation dataset (by omitting the line dataset[name] = dataset[name].batch(batch_size) in data_fn) I get the same error but after a much longer time.

I can only avoid this error if I don't batch the data and use steps=1 for evaluation (does that perform the evaluation on the whole dataset?)

I don't understand what causes this error as the documentation suggests I should be able to evaluate on batches too.

Note: I get the same error when using tf.estimator.evaluate on data batches.

### Source code / logs

```
  File "/usr/lib/python2.7/site-packages/tensorflow/python/estimator/training.py", line 439, in train_and_evaluate
    executor.run()
  File "/usr/lib/python2.7/site-packages/tensorflow/python/estimator/training.py", line 518, in run
    self.run_local()
  File "/usr/lib/python2.7/site-packages/tensorflow/python/estimator/training.py", line 657, in run_local
    eval_result = evaluator.evaluate_and_export()
  File "/usr/lib/python2.7/site-packages/tensorflow/python/estimator/training.py", line 847, in evaluate_and_export
    hooks=self._eval_spec.hooks)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 425, in evaluate
    name=name)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 1085, in _evaluate_model
    input_fn, model_fn_lib.ModeKeys.EVAL))
  File "/usr/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 691, in _get_features_and_labels_from_input_fn
    result = self._call_input_fn(input_fn, mode)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 798, in _call_input_fn
    return input_fn(**kwargs)
  File "/snel/home/mreza/projects/PBT_HP_opt/lfadslite/lfadslite.py", line 748, in <lambda>
    eval_input = lambda: self.data_fn(hps, tf.estimator.ModeKeys.EVAL)
  File "/snel/home/mreza/projects/PBT_HP_opt/lfadslite/lfadslite.py", line 112, in data_fn
    next_element = iterator.get_next()
  File "/usr/lib/python2.7/site-packages/tensorflow/python/data/ops/iterator_ops.py", line 370, in get_next
    name=name)), self._output_types,
  File "/usr/lib/python2.7/site-packages/tensorflow/python/ops/gen_dataset_ops.py", line 1466, in iterator_get_next
    output_shapes=output_shapes, name=name)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 3392, in create_op
    op_def=op_def)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1718, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

OutOfRangeError (see above for traceback): End of sequence
	 [[Node: IteratorGetNext = IteratorGetNext[output_shapes=[[?,200,29]], output_types=[DT_FLOAT], _device="/job:localhost/replica:0/task:0/device:CPU:0"](OneShotIterator)]]

```

**Comment by mrry** [comment]
@xiejw To my eye, it looks like @mrezak is doing everything correctly here (setting `EvalSpec(..., steps=None, ...)` and returning a finite iterator from the input function). Is there something missing?

**Comment by xiejw** [comment]
If I understand correctly, this issue is "once give estimator an input_fn with dataset inside, the evaluate process will error out with OutOfRangeError."

Estimator can handle this correctly actually. However, a known common root cause for this is metrics defined in model_fn have bug. We need to rule that part out first. 

@mrezak if possible, can you show the code about the model_fn? Or if you have a minimal reproducible script, that will be extremely helpful. -- Thanks in advance. 

A common problem for this is: metric in tensorflow should return two Ops: update_op and value_op. Estimator calls the update_op for each batch of the data in input source and, once it is exhausted, it call the value_op to get the metric values. The value_op here should have dependency back to variables reading only. 

Many model_fn puts the dependency of value_op with the input pipeline, so, estimator.evaluate will thereby trigger the input pipeline one more time, which errors out with OutOfRangeError

**Comment by mrezak** [comment]
Thanks @mrry and @xiejw for looking into this. @xiejw I see! Actually my value_op depended on the input pipeline (returning different form of loss), so as you pointed out that is the cause of this error. 

I have written minimal example script (A simple autoencoder) that reproduces the EndOfSequence error (If I replace `dummy_eval` with a tensor that does not depend on the input pipeline, I won't get the EndOfSequence Error) :

```
import numpy as np
import tensorflow as tf

def data_fn(data_dict, batch_size, mode, num_epochs=10):
    if mode == tf.estimator.ModeKeys.TRAIN:
        dataset = tf.data.Dataset.from_tensor_slices(data_dict['train_data'].astype(np.float32))
        dataset = dataset.cache()
        dataset = dataset.shuffle(buffer_size= batch_size * 10).repeat(num_epochs).batch(batch_size)
    else:
        dataset = tf.data.Dataset.from_tensor_slices(data_dict['valid_data'].astype(np.float32))
        dataset = dataset.cache()
        dataset = dataset.batch(batch_size)

    iterator = dataset.make_one_shot_iterator()
    next_element = iterator.get_next()

    return next_element, tf.convert_to_tensor('dummy_dataname')

def model_fn(features, labels, mode, params):
    with tf.variable_scope('encoder'):
        cell = tf.contrib.rnn.GRUCell(num_units=20)
        _, output_latent = tf.nn.dynamic_rnn(cell=cell, inputs=features, dtype=tf.float32)
    with tf.variable_scope('decoder'):
        cell = tf.contrib.rnn.GRUCell(num_units=20)
        ts = tf.shape(features)
        #ts = tf.Print(ts, [ts])
        z_inps = tf.zeros([ts[0], ts[1], 1])
        output_recon, _ = tf.nn.dynamic_rnn(cell=cell, inputs=z_inps, initial_state=output_latent, dtype=tf.float32)    

    output_recon = tf.contrib.layers.fully_connected(inputs=output_recon, num_outputs=params['dims'], activation_fn=None)
    vars   = tf.trainable_variables() 
    dummy_lossL2 = tf.add_n([ tf.nn.l2_loss(v) for v in vars ]) * 0.001
    recon_loss = tf.losses.mean_squared_error(features, output_recon)
    loss =  recon_loss + 0.1 * dummy_lossL2
    dummy_eval = recon_loss
    global_step = tf.train.get_global_step()
    train_op = tf.train.AdamOptimizer(0.01).minimize(loss, global_step)
    eval_metric_ops = {'dummy': (dummy_eval, dummy_eval)}
    logging_hook = tf.train.LoggingTensorHook({'step':global_step, 'loss':loss}, every_n_iter=1)
    return tf.estimator.EstimatorSpec(mode=mode,
                                  loss=loss,
                                  train_op=train_op,
                                  eval_metric_ops=eval_metric_ops,
                                  training_hooks=[logging_hook]
                                  )

def train_model():
    tf.logging.set_verbosity(tf.logging.INFO)
    config = tf.ConfigProto(allow_soft_placement=True,
                            log_device_placement=False)
    config.gpu_options.allow_growth = True
    run_config = tf.contrib.learn.RunConfig(
        save_checkpoints_steps=100,
        keep_checkpoint_max=10,
        session_config=config
    )
    hps = {'dims': 5}
    
    # make training and validation data: sinusoids with random phases
    num_samp_tr = 1000
    num_samp_val = 1000
    ramps_tr = np.transpose(np.broadcast_to(0.1*np.arange(0,100), (num_samp_tr, hps['dims'], 100)), (0, 2, 1))
    rand_phase = np.transpose(np.tile(np.random.randn(num_samp_tr, hps['dims']), (100,1,1)), (1, 0, 2))
    ramps_val = np.transpose(np.broadcast_to(0.1*np.arange(0,100), (num_samp_val, hps['dims'], 100)), (0, 2, 1))
    rand_phase_val = np.transpose(np.tile(np.random.randn(num_samp_val, hps['dims']), (100,1,1)), (1, 0, 2))
    data = {'train_data': np.sin(ramps_tr + rand_phase),
            'valid_data': np.sin(ramps_val + rand_phase_val)}

    train_input = lambda: data_fn(data, 100, tf.estimator.ModeKeys.TRAIN, num_epochs=10)
    eval_input = lambda: data_fn(data, 1000, tf.estimator.ModeKeys.EVAL)
    estimator = tf.estimator.Estimator(model_fn=model_fn, params=hps, config=run_config)
    train_spec = tf.estimator.TrainSpec(train_input, max_steps=100)
    eval_spec = tf.estimator.EvalSpec(eval_input,
                                      steps=None,
                                      throttle_secs = 100)

    tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)

train_model()
```

The problem in my actual code is that my total loss to be optimized is composed of multiple losses (reconstruction + L2 + KL) and in the evaluation part I want to get the reconstruction loss (on the validation data), which depend on the input data pipeline. My actual reconstruction cost is more complex than MSE (none of the other tf.metric functions as well), so in this case do I have to construct my reconstruction cost calculation using tf.metric basic functions to be able to use it as eval_metric_ops?

**Comment by xiejw** [comment]
Hi @mrezak 

Does the following approach work for you

    my_total_loss =  ... # the loss you care. Pay attention to how you reduce the loss. 
    eval_metric_ops = {'total_loss: tf.metrics.mean(my_total_loss)}

This should give you the loss on held-out data.

**Comment by mrezak** [comment]
Thanks very much @xiejw. This seems to work for me.

**Comment by xiejw** [comment]
Sure. I will thereby close this issue. I will try to improve the error message you saw and give a clear actionable step for user when hitting this issue.

**Comment by mukurgupta** [comment]
@xiejw I have a similar code structure and I am facing the same issue as well. I tried what you said above but it didn't work.
